Source/Core/Core/LibusbUtils.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <mutex>
+#include <thread>
+
+#if defined(__LIBUSB__)
+#include <libusb.h>
+#endif
+
+#include "Common/Assert.h"
+#include "Common/Flag.h"
+#include "Common/Thread.h"
+#include "Core/LibusbUtils.h"
+
+namespace LibusbUtils
+{
+#if defined(__LIBUSB__)
+class Context::Impl
+{
+public:
+  Impl()
+  {
+    const int ret = libusb_init(&m_context);
+    ASSERT_MSG(IOS_USB, ret == LIBUSB_SUCCESS, "Failed to init libusb: %s", libusb_error_name(ret));
+    if (ret != LIBUSB_SUCCESS)
+      return;
+
+#ifdef _WIN32
+    libusb_set_option(m_context, LIBUSB_OPTION_USE_USBDK);
+#endif
+    m_event_thread_running.Set();
+    m_event_thread = std::thread(&Impl::EventThread, this);
+  }
+
+  ~Impl()
+  {
+    if (!m_context || !m_event_thread_running.TestAndClear())
+      return;
+    libusb_exit(m_context);
```

Comment:
```suggestion
    libusb_interrupt_event_handler(m_context);
    libusb_exit(m_context);
```

Fixes the hang at shutdown on macOS. Maybe the `libusb_exit` call should be moved after the `join`? Seems to work either way…

Reply:
Thanks for investigating the issue! I've added the call to libusb_interrupt_event_handler.

Strangely enough libusb_interrupt_event_handler isn't mentioned in the online documentation.

Reply:
Apparently FreeBSD's reimplementation of libusb is too old and does not provide this function… So I've changed the event handling timeout to 5 seconds. There should be no effect on other platforms, but will prevent hanging on close on FreeBSD. Can't do any better than that.